libsolidity/ast/Types.cpp

```diff
@@ -566,12 +575,40 @@ BoolResult IntegerType::isImplicitlyConvertibleTo(Type const& _convertTo) const
 
 BoolResult IntegerType::isExplicitlyConvertibleTo(Type const& _convertTo) const
 {
-	return _convertTo.category() == category() ||
-		_convertTo.category() == Category::Address ||
-		_convertTo.category() == Category::Contract ||
-		_convertTo.category() == Category::Enum ||
-		(_convertTo.category() == Category::FixedBytes && numBits() == dynamic_cast<FixedBytesType const&>(_convertTo).numBytes() * 8) ||
-		_convertTo.category() == Category::FixedPoint;
+	if (auto integerType = dynamic_cast<IntegerType const*>(&_convertTo))
+	{
+		if (
+			(numBits() == integerType->numBits()) ||
+			(isSigned() == integerType->isSigned())
+		)
+			return true;
+	}
+	else if (_convertTo.category() == Category::Address)
+	{
+		if (!isSigned() && numBits() == 160)
+			return true;
+	}
+	else if (auto fixedBytesType = dynamic_cast<FixedBytesType const*>(&_convertTo))
+	{
+		if (!isSigned() && (numBits() == fixedBytesType->numBytes() * 8))
```

Comment:
Some of the tests such as this and the one above are repeated in some places. I'm wondering it it'd make sense to extract them somehow. Otherwise this looks good too.

Reply:
I thought about extracting into something like `canExplicitlyConvert(Type& _a, Type& _b)` and do it there. Maybe I'll do it after fixing the tests.

Reply:
I was thinking about converting each of these comparisons into individual functions, for example `canConvert(IntegerType const& _a, FixedBytesType const& _b)`. Since the rules for converting `a -> b` and `b -> a` are the same, (provided both conversions are feasible) both the cases can call the same function, except in one case, the arguments will have to be flipped. So these checks are not repeated. But I couldn't find a nice way to implement it. So I'll leave it like this.

Reply:
Ok, I think that's fine too

Reply:
Some way to make symmetry explicit would be nice, but I also don't see an easy way to do it.